Для "мешка слов" был отобран набор отзывов на наушники X (название скрыто ибо за рекламу мне не заплатили :), купленные мной на маркетплейсе Y. (здесь тоже самое по понятным причинам ;)

In [1]:
# Создание списка уникальных слов

feedback_list = [
    '''Купила наушники недавно, и пока очень довольна. Звук чистый, басы глубокие, хотя иногда верхние частоты слегка режут ухо. Удобно сидят на голове, даже при длительном использовании не вызывают дискомфорта. Подсветка добавляет стильности, хотя ночью мешает спать.''',

    '''Не ожидал такого качества за эти деньги. Микрофон отлично передает звук, а амбушюры мягкие и удобные. Единственное, что кабель кажется немного хлипким.''',

    '''Наушники очень удобные, легкие, но звук оставляет желать лучшего. При повышении громкости появляются искажения. Для фильмов и разговоров годятся, но для музыки советую что-то получше.''',

    '''Использую эти наушники второй месяц. Пока все устраивает. Немного давят на голову, но это терпимо. Звук объемный, особенно радует в играх. Однако с музыкой есть проблемы – басы слишком глухие.''',

    '''Ожидал большего. Подсветка работает исправно, но микрофон уже через неделю начал фонировать. Сам звук средний: нет глубины и объемности. Наушники выглядят красиво, но качество сборки под вопросом.''',

    '''Покупал наушники для работы и общения. Для этих целей они идеальны. Легкие, хорошо передают речь, удобный регулятор громкости на проводе. Для музыки – не лучший выбор, но меня устраивает.''',

    '''Наушники отличные, хотя уши устают после нескольких часов. Амбушюры мягкие, но чуть маловаты. Звук хороший, с объемным эффектом. Подсветка очень яркая, но отключить её нельзя, что немного раздражает.''',

    '''Получил наушники быстро. Качество сборки на высоте, звук радует – есть и басы, и четкость высоких частот. Минус – кабель кажется слишком жестким.''',

    '''Взял для подарка. Наушники легкие, звук чистый, хотя немного не хватает низких частот. Понравилось, что можно отключить подсветку. Отличный выбор за свою цену.''',

    '''Средние наушники. Звук хороший только на средней громкости, при увеличении начинают хрипеть. Провод хлипкий, но микрофон порадовал – четкий звук без лишних шумов.''',
]


print(f'Количество отзывов: {len(feedback_list)}')

Количество отзывов: 10


In [2]:
# Выявление наиболее популярных слов, чтобы исключить из них возможные стоп-слова
tokens_dict = {}

for review in feedback_list:
    for word in review.split():
        word = word.lower()
        if word in tokens_dict:
            tokens_dict[word] += 1
        else:
            tokens_dict[word] = 1
            
sorted_tokens = sorted(tokens_dict.items(), key=lambda item: item[1], reverse=True)

for word, frequency in sorted_tokens:
    print(f"{word}: {frequency}")

звук: 9
но: 9
наушники: 8
и: 7
для: 6
на: 5
–: 5
хотя: 4
не: 4
немного: 4
очень: 3
при: 3
подсветка: 3
микрофон: 3
что: 3
легкие,: 3
пока: 2
чистый,: 2
басы: 2
ожидал: 2
за: 2
эти: 2
амбушюры: 2
кабель: 2
кажется: 2
громкости: 2
музыки: 2
устраивает.: 2
радует: 2
с: 2
есть: 2
слишком: 2
качество: 2
сборки: 2
отключить: 2
частот.: 2
купила: 1
недавно,: 1
довольна.: 1
глубокие,: 1
иногда: 1
верхние: 1
частоты: 1
слегка: 1
режут: 1
ухо.: 1
удобно: 1
сидят: 1
голове,: 1
даже: 1
длительном: 1
использовании: 1
вызывают: 1
дискомфорта.: 1
добавляет: 1
стильности,: 1
ночью: 1
мешает: 1
спать.: 1
такого: 1
качества: 1
деньги.: 1
отлично: 1
передает: 1
звук,: 1
а: 1
мягкие: 1
удобные.: 1
единственное,: 1
хлипким.: 1
удобные,: 1
оставляет: 1
желать: 1
лучшего.: 1
повышении: 1
появляются: 1
искажения.: 1
фильмов: 1
разговоров: 1
годятся,: 1
советую: 1
что-то: 1
получше.: 1
использую: 1
второй: 1
месяц.: 1
все: 1
давят: 1
голову,: 1
это: 1
терпимо.: 1
объемный,: 1
особенно: 1
в: 1
играх.: 1
однако:

In [3]:
# Составление словаря стоп-слов
stop_tokens = [
    "не", "в", "но", "и", "на", "через", "что", "с", "а", "для", "от", 
    "у", "меня", "все", "это", "как", "по", "если", "про", "мне", "иногда", 
    "надо", "чуть", "после", "ещё", "плохо", "сами", "чуть", "какие-то", 
    "или", "есть", "всё", "при", "в", "их", "можно", "при", "до", "был", 
    "что", "везде", "где", "немного", "так", "эти", "все"
]

print(f'Количество стоп-слов: {len(stop_tokens)}')

Количество стоп-слов: 46


In [4]:
# Масштабирование данных с помощью tf-idf
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vectorizer = TfidfVectorizer(stop_words=stop_tokens)
X_tfidf = tfidf_vectorizer.fit_transform(feedback_list)

print("TF-IDF матрица:", X_tfidf.toarray())

TF-IDF матрица: [[0.         0.14145064 0.         ... 0.         0.         0.        ]
 [0.23783351 0.         0.         ... 0.         0.         0.        ]
 [0.         0.         0.         ... 0.         0.         0.        ]
 ...
 [0.         0.20675349 0.         ... 0.         0.         0.        ]
 [0.         0.         0.         ... 0.         0.         0.        ]
 [0.         0.         0.25278913 ... 0.         0.         0.        ]]


/home/baitemir/projects/big data/lab7/venv/lib/python3.12/site-packages/sklearn/feature_extraction/text.py:402: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['какие', 'то'] not in stop_words.
  warnings.warn(


In [5]:
# Использование n-gram (1, 2), pipeline(vectorizer, regressor)
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.feature_extraction.text import TfidfVectorizer

# Пример меток для отзывов: 0 - негативный, 1 - положительный
labels = [1, 1, 0, 1, 1, 0, 1, 1, 0, 1]  # Пример меток для 10 отзывов
  # Должно быть 25 меток

# Разделение данных на обучающую и тестовую выборки
X_train, X_test, y_train, y_test = train_test_split(feedback_list, labels, test_size=0.2, random_state=42)

# Определение пайплайна с векторизатором и логистической регрессией
pipeline = Pipeline([
    ('vectorizer', TfidfVectorizer(ngram_range=(1, 2), stop_words=['и', 'в', 'на', 'с', 'не', 'но', 'для', 'что', 'это'])),
    ('classifier', LogisticRegression())
])

# Обучение модели
pipeline.fit(X_train, y_train)

# Предсказания на тестовых данных
y_pred = pipeline.predict(X_test)

# Оценка точности модели
print("Точность модели:", accuracy_score(y_test, y_pred))



Точность модели: 0.5


In [6]:
# Исследование метрик модели
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

print("Точность модели:", accuracy_score(y_test, y_pred))

print("\nОтчет по меткам (precision, recall, f1-score):")
print(classification_report(y_test, y_pred))

print("\nМатрица ошибок:")
print(confusion_matrix(y_test, y_pred))

print("\nРеальные метки vs предсказанные метки:")
for true, pred in zip(y_test, y_pred):
    print(f"Реальная метка: {true}, Предсказанная метка: {pred}")

Точность модели: 0.5

Отчет по меткам (precision, recall, f1-score):
              precision    recall  f1-score   support

           0       0.00      0.00      0.00         1
           1       0.50      1.00      0.67         1

    accuracy                           0.50         2
   macro avg       0.25      0.50      0.33         2
weighted avg       0.25      0.50      0.33         2


Матрица ошибок:
[[0 1]
 [0 1]]

Реальные метки vs предсказанные метки:
Реальная метка: 0, Предсказанная метка: 1
Реальная метка: 1, Предсказанная метка: 1


/home/baitemir/projects/big data/lab7/venv/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/home/baitemir/projects/big data/lab7/venv/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/home/baitemir/projects/big data/lab7/venv/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modif

In [7]:
# Вывод слов с наибольшим вкладом
model = pipeline.named_steps['classifier']
coefficients = model.coef_

tokens = pipeline.named_steps['vectorizer'].get_feature_names_out()
top_10_tokens = sorted(zip(coefficients[0], tokens), reverse=True)[:10]
print("Топ-10 слов с наибольшим вкладом:")

for word in top_10_tokens:
    print(word)

Топ-10 слов с наибольшим вкладом:
(np.float64(0.07041380892458138), 'хотя')
(np.float64(0.06828194196557347), 'басы')
(np.float64(0.06420553793489321), 'подсветка')
(np.float64(0.05706637682034539), 'слишком')
(np.float64(0.05706637682034539), 'радует')
(np.float64(0.05706637682034539), 'есть')
(np.float64(0.05674995216282369), 'сборки')
(np.float64(0.05674995216282369), 'качество сборки')
(np.float64(0.05674995216282369), 'качество')
(np.float64(0.05601170366595769), 'хороший')


### **Вывод:**

В данной лабораторной работе был использован метод **"мешок слов"** для анализа отзывов на наушники.

#### Результаты

---

**Точность модели**: 0.5

**Отчет по меткам (precision, recall, f1-score)**:

|       | precision | recall | f1-score | support |
|-------|-----------|--------|----------|---------|
| 0     | 0.00      | 0.00   | 0.00     | 1       |
| 1     | 0.50      | 1.00   | 0.67     | 1       |

**Общая точность**: 0.50

**Средние значения**:

|       | precision | recall | f1-score |
|-------|-----------|--------|----------|
| macro avg | 0.25      | 0.50   | 0.33     |
| weighted avg | 0.25      | 0.50   | 0.33     |

**Матрица ошибок**:
```
[[0 1]  
 [0 1]]
```

**Реальные метки vs предсказанные метки**:  
Реальная метка: 0, Предсказанная метка: 1  
Реальная метка: 1, Предсказанная метка: 1

Из-за малого объема данных модель не показала высокую точность. Некоторые метки были неправильно предсказаны (например, метка 0 была ошибочно предсказана как 1). Также модель имеет трудности с определением точности для класса с отсутствующими предсказаниями, что вызвало предупреждения.

**Топ-10 слов с наибольшим вкладом**:
- Хотя: 0.0704
- Басы: 0.0683
- Подсветка: 0.0642
- Слишком: 0.0571
- Радует: 0.0571
- Есть: 0.0571
- Сборки: 0.0567
- Качество сборки: 0.0567
- Качество: 0.0567
- Хороший: 0.0560

Матрица признаков TF-IDF показывает, какие слова в отзывах оказались наиболее важными для модели. Наибольший вклад внесли слова, связанные с качеством звука, сборкой и комфортом наушников.